### PR TITLE
Rustup

### PR DESCRIPTION
**Pull request by RalfJung**
Updates for https://github.com/rust-lang/rust/pull/53779

Functional changes:
* Restrict what pointer equality can do, to make sure it agrees with run-time behavior. For now, only comparing within the same allocation, comparing two life pointers (i.e. allocated and in-bounds), and comparing a pointer with NULL are allowed.

### CODE REVIEW

**Comment by RalfJung** [comment]
I had to make the pointer equality a bit smarter: When comparing a really big integer with a pointer that cannot possibly be so big or else the allocation would wrap around the end of the address space, we also say these are inequal. This is needed for `Rc` (which compares with `usize::MAX`).

File: src/operator.rs [hunk 1]
```diff
@@ -148,11 +148,30 @@ impl<'a, 'mir, 'tcx> EvalContextExt<'tcx> for EvalContext<'a, 'mir, 'tcx, super:
...
+            (Scalar::Bits { bits, size }, Scalar::Ptr(ptr)) => {
+                assert_eq!(size as u64, self.pointer_size().bytes());
+                if bits == 0 {
+                    // Nothing equals 0
+                    false
+                } else {
+                    // Compute the highest address at which this allocation could live
+                    let alloc = self.memory.get(ptr.alloc_id)?;
+                    let max_base_addr =
+                        (1u128 << self.pointer_size().bits()) - alloc.bytes.len() as u128;
```
Comment (oli-obk) [inline src/operator.rs]: for alignment > 1 this could be even more permissive, right?

File: src/operator.rs [hunk 2]
```diff
@@ -148,11 +148,30 @@ impl<'a, 'mir, 'tcx> EvalContextExt<'tcx> for EvalContext<'a, 'mir, 'tcx, super:
...
+                    let max_addr = max_base_addr + ptr.offset.bytes() as u128;
+                    if bits > max_addr {
+                        // The integer is too big, this cannot possibly be equal
+                        false
+                    } else {
+                        // TODO: We could also take alignment into account
```
Comment (oli-obk) [inline src/operator.rs]: oh, you already added a todo, ignore my above comment

**Comment by RalfJung** [comment]
Since you clearly took a look and had no reservations, I assume I can merge once CI is green?

**Pull request merged**